File: leetcode/450.删除二叉搜索树中的节点/450.删除二叉搜索树中的节点.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        # base case
        if root is None:
            return root
        if root.val == key:
            # 如果不存在左节点，返回右节点
            # 如果不存在右节点，返回左节点
            # 两个不存在也是返回的None
            if root.left is None:
                return root.right
            if root.right is None:
                return root.left
            # 处理 左右子树都存在，查找右子树最小并对值进行替换
            min_node = self.findMin(root.right)
            root.val = min_node.val
            # 删除替换的最小节点
            root.right = self.deleteNode(root.right, min_node.val)
        elif root.val > key:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)

        return root

    def findMin(self, node):
        while node.left:
            node = node.left
        return node

File: leetcode/450.删除二叉搜索树中的节点/450.删除二叉搜索树中的节点.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil{
		return root
	}
	//找到值
	if root.Val == key{
		// 处理有一个子树，或者无子树时情况
		if root.Left == nil{
			return root.Right
		}
		if root.Right == nil{
			return root.Left
		}
		// 有双节点时，找到右子树最小的值，替换值，删除最小值的节点
		minNode := findMin(root.Right)
		root.Val = minNode.Val
		root.Right = deleteNode(root.Right, minNode.Val)
	}else if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	}else if root.Val < key{
		root.Right = deleteNode(root.Right, key)
	}
	return root
}
func findMin(node *TreeNode) *TreeNode{
	for node.Left != nil{
		node = node.Left
	}
	return node
}
